pyccel/parser/semantic.py

```diff
@@ -2698,17 +2698,35 @@ def _handle_is_operator(self, IsClass, expr, **settings):
             var1, var2 = var2, var1
 
         if isinstance(var2, Nil):
+            if isinstance(var1, Nil):
+                if IsClass == IsNot:
+                    return BooleanFalse()
+                elif IsClass == Is:
+                    return BooleanTrue()
             if not var1.is_optional:
                 errors.report(PYCCEL_RESTRICTION_OPTIONAL_NONE,
                         bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
                         severity='error', blocker=self.blocking)
             return IsClass(var1, expr.rhs)
 
+        if (var1.dtype != var2.dtype):
+            if IsClass == Is:
+                return BooleanFalse()
+            elif IsClass == IsNot:
+                return BooleanTrue()
+
         if ((var1.is_Boolean or isinstance(var1.dtype, NativeBool)) and
             (var2.is_Boolean or isinstance(var2.dtype, NativeBool))):
             return IsClass(var1, var2)
 
-        errors.report(PYCCEL_RESTRICTION_IS_RHS,
+        lst = ["string", "complex", "real", "int"]
```

Comment:
Please use `NativeString()`, `NativeComplex()`, `NativeReal()`, `NativeInt()` instead of converting to strings